### PR TITLE
Implementation change for `search` in graph interface

### PR DESCRIPTION
**Pull request by ilaner**
## Related Issues
fixes: link to the issue

## Description
1. Load models which does not have relationships
2. Use multiprocessing in model loading
3. Fix relationship data hash function
4. Normalization name fixes 
## Screenshots
Paste here any images that will help the reviewer

## Must have
- [ ] Tests
- [ ] Documentation
- [ ] [VS Code Extension](https://docs.gitlab.com/ee/ci/yaml/#tags)
  - [ ] Functionality implemented in the extension - <u>\<link-to-pr-here\></u>
  - [ ] N/A - Functionality cannot be implemented in the extension because <u>\<add-reason-here\></u>

### CODE REVIEW

File: demisto_sdk/commands/content_graph/interface/neo4j/neo4j_graph.py [hunk 1]
```diff
@@ -165,20 +153,16 @@ def _add_relationships(
...
+            nodes_to.extend(pack.nodes_to)
+        self._add_nodes_to_mapping(nodes_to)
 
-        for pack in mandatorily_dependencies:
-            obj = Neo4jContentGraphInterface._id_to_obj[pack.node_from.id]
+        for id, pack in mandatorily_dependencies.items():
+            obj = Neo4jContentGraphInterface._id_to_obj[id]
             for node_to in pack.nodes_to:
                 target = Neo4jContentGraphInterface._id_to_obj[node_to.id]
                 obj.relationships_data[RelationshipType.DEPENDS_ON].add(
```
Comment (dantavori) [inline demisto_sdk/commands/content_graph/interface/neo4j/neo4j_graph.py:168]: can we do an inner method for this?

**Review by dantavori — COMMENTED** [review]
1. docstrings
2. add_relationships_data - seperate to functions
3. where join - fix as discussed

File: demisto_sdk/commands/content_graph/tests/update_content_graph_test.py [hunk 2]
```diff
@@ -423,7 +424,7 @@ def test_update_content_graph(
                 content_type=ContentType.PACK,
                 all_level_dependencies=True,
             )
-
+            print("packs_from_graph", packs_from_graph)
```
Comment (dantavori) [inline demisto_sdk/commands/content_graph/tests/update_content_graph_test.py:427]:
```suggestion
```

File: demisto_sdk/commands/content_graph/objects/pack.py [hunk 3]
```diff
@@ -100,6 +100,9 @@ def __iter__(self) -> Generator[ContentItem, Any, Any]:  # type: ignore
         for content_items in vars(self).values():
             yield from content_items
 
+    def __bool__(self) -> bool:
+        return bool(list(self))
```
Comment (dantavori) [inline demisto_sdk/commands/content_graph/objects/pack.py:104]:
```suggestion
    def __bool__(self) -> bool:
        """Used for easier determination of content items existence in a pack. """
        return bool(list(self))
```

File: demisto_sdk/commands/content_graph/objects/base_content.py [hunk 4]
```diff
@@ -127,8 +127,11 @@ def from_path(path: Path) -> Optional["BaseContent"]:
     def dump(self, path: DirectoryPath, marketplace: MarketplaceVersions) -> None:
         pass
 
+    def add_relationship(self, relationship_type: RelationshipType, relationship: "RelationshipData") -> None:
+        self.relationships_data[relationship_type].add(relationship)
+
 
-class ServerContent(BaseContent):
+class UnknownContent(BaseContent):
```
Comment (dantavori) [inline demisto_sdk/commands/content_graph/objects/base_content.py:134]:
```suggestion
class UnknownContent(BaseContent):
    """A model for non-existing content items used by existing content items. """
```

File: demisto_sdk/commands/content_graph/interface/neo4j/queries/relationships.py [hunk 5]
```diff
@@ -189,3 +190,24 @@ def create_relationships_by_type(
     result = run_query(tx, query, data=data).single()
     merged_relationships_count: int = result["relationships_merged"]
     logger.info(f"Merged {merged_relationships_count} relationships of type {relationship}.")
+
+
+def _match_relationships(
```
Comment (dantavori) [inline demisto_sdk/commands/content_graph/interface/neo4j/queries/relationships.py:195]: please add a docstring

File: demisto_sdk/commands/content_graph/interface/neo4j/queries/nodes.py [hunk 6]
```diff
@@ -122,31 +122,31 @@ def _match(
     tx: Transaction,
     marketplace: MarketplaceVersions = None,
     content_type: Optional[ContentType] = None,
-    filter_list: Optional[Iterable[int]] = None,
+    ids_list: Optional[Iterable[int]] = None,
     **properties,
-) -> List[Neo4jResult]:
+) -> List[Neo4jRelationshipResult]:
```
Comment (dantavori) [inline demisto_sdk/commands/content_graph/interface/neo4j/queries/nodes.py:127]: same here

File: demisto_sdk/commands/content_graph/interface/neo4j/queries/dependencies.py [hunk 7]
```diff
@@ -148,16 +148,21 @@ def create_depends_on_relationships(tx: Transaction) -> None:
         AND NOT pack_b.name IN {IGNORED_PACKS_IN_DEPENDENCY_CALC}
         AND a.is_test <> true
         AND b.is_test <> true
-        WITH r, pack_a, pack_b
+        WITH r, a, b, pack_a, pack_b
```
Comment (dantavori) [inline demisto_sdk/commands/content_graph/interface/neo4j/queries/dependencies.py:151]: DMd you a fix for this one

File: demisto_sdk/commands/content_graph/interface/neo4j/queries/dependencies.py [hunk 8]
```diff
@@ -4,13 +4,11 @@
...
 from demisto_sdk.commands.common.constants import GENERIC_COMMANDS_NAMES, REPUTATION_COMMAND_NAMES, MarketplaceVersions
-from demisto_sdk.commands.content_graph.common import ContentType, Neo4jResult, RelationshipType
+from demisto_sdk.commands.content_graph.common import ContentType, Neo4jRelationshipResult, RelationshipType
 from demisto_sdk.commands.content_graph.interface.neo4j.queries.common import (is_target_available, run_query,
                                                                                to_neo4j_map)
 
-REPUTATION_COMMANDS_NODE_IDS = [
-    f"{ContentType.COMMAND}:{cmd}" for cmd in REPUTATION_COMMAND_NAMES
-]
+REPUTATION_COMMANDS_NODE_IDS = [f"{ContentType.COMMAND}:{cmd}" for cmd in REPUTATION_COMMAND_NAMES]
```
Comment (dantavori) [inline demisto_sdk/commands/content_graph/interface/neo4j/queries/dependencies.py:11]: do we use these constants?

File: demisto_sdk/commands/content_graph/interface/neo4j/neo4j_graph.py [hunk 9]
```diff
@@ -24,15 +25,39 @@
...
+    Raises:
+        NoModelException: If no model found to parse on
+    """
+    obj: BaseContent
+    content_type = node.get("content_type", "")
+    if node.get("not_in_repository"):
+        obj = UnknownContent.parse_obj(node)
+
+    else:
+        model = content_type_to_model.get(content_type)
```
Comment (dantavori) [inline demisto_sdk/commands/content_graph/interface/neo4j/neo4j_graph.py:53]: interesting that it works when content_type_to_model keys are enums

**Review by dantavori — APPROVED** [review]
please fix notes before merging

File: demisto_sdk/commands/content_graph/interface/neo4j/queries/dependencies.py (hunk 8 above)
Comment (ilaner) [reply demisto_sdk/commands/content_graph/interface/neo4j/queries/dependencies.py:11]: no, deleting

**Review by dantavori — APPROVED** [review]

File: demisto_sdk/commands/content_graph/interface/neo4j/queries/dependencies.py [hunk 10]
```diff
@@ -113,25 +125,51 @@ def update_marketplaces_property(tx: Transaction, marketplace: str) -> None:
...
 
-    WITH count(rcmd) as command_count, content_item, r, integration
     MERGE (content_item)-[u:USES]->(integration)
-    SET u.mandatorily = u.mandatorily AND (CASE WHEN command_count = 1 THEN true ELSE false END)
+    ON CREATE
+        SET u.mandatorily = CASE WHEN command_count = 1 THEN r.mandatorily ELSE false END
+    ON MATCH
+        SET u.mandatorily = u.mandatorily OR (CASE WHEN command_count = 1 THEN r.mandatorily ELSE false END)
     RETURN count(u) as uses_relationships
+
```
Comment (dantavori) [inline demisto_sdk/commands/content_graph/interface/neo4j/queries/dependencies.py:176]: let's log the query outputs (please check)
```suggestion
        SET u.mandatorily = u.mandatorily OR (CASE WHEN command_count = 1 THEN r.mandatorily ELSE false END)
    RETURN
        content_item.node_id AS content_item,
        r.mandatorily AS is_cmd_mandatory,
        collect(integration.object_id) AS integrations,
        u.mandatorily AS is_integ_mandatory,
        command.name AS command
"""
    result = run_query(tx, query)
    for row in result:
        content_item = row["content_item"]
        command = row["command"]
        is_cmd_mandatory = "mandatory" if row["is_cmd_mandatory"] else "optional"
        integrations = row["integrations"]
        is_integ_mandatory = "mandatory" if row["is_integ_mandatory"] else "optional"
        content_item = row["content_item"]
        msg = f"{content_item} uses command {command} ({is_cmd_mandatory}), "
        msg += f"new {is_integ_mandatory} rels to integrations: {integrations}"
        logger.debug(msg)


```

File: demisto_sdk/commands/content_graph/parsers/layout.py [hunk 11]
```diff
@@ -49,7 +49,7 @@ def connect_to_dependencies(self) -> None:
             )
 
         for field in self.get_field_ids_recursively():
-            self.add_dependency_by_id(field, dependency_field_type, is_mandatory=False)
+            self.add_dependency_by_id(field, dependency_field_type, is_mandatory=False)  # TODO: Check if mandatory
```
Comment (dantavori) [inline demisto_sdk/commands/content_graph/parsers/layout.py:52]:
```suggestion
            self.add_dependency_by_id(field, dependency_field_type, is_mandatory=False)
```

File: demisto_sdk/commands/content_graph/objects/pack.py [hunk 12]
```diff
@@ -243,6 +250,15 @@ def dump(self, path: Path, marketplace: MarketplaceVersions):
                 shutil.copy(self.path / "Author_image.png", path / "Author_image.png")
             except FileNotFoundError:
                 logger.info(f'No such file {self.path / "Author_image.png"}')
+            if self.object_id == "Base":
```
Comment (dantavori) [inline demisto_sdk/commands/content_graph/objects/pack.py:253]: please move to sep method handle_base_pack_documentation_files() and use constants

File: demisto_sdk/commands/content_graph/interface/neo4j/queries/dependencies.py [hunk 13]
```diff
@@ -1,59 +1,74 @@
...
     remove_existing_depends_on_relationships(tx)
     fix_marketplaces_properties(tx)
     update_uses_for_integration_commands(tx)
+    delete_deprecatedcontent_relationship(tx)  # TODO decide what to do with this
     create_depends_on_relationships(tx)
 
 
+def delete_deprecatedcontent_relationship(tx: Transaction) -> None:
+    query = f"""
+        MATCH (source) - [r:{RelationshipType.USES}] -> (target) - [:{RelationshipType.IN_PACK}] -> (:{ContentType.PACK}{{object_id: "DeprecatedContent"}})
```
Comment (dantavori) [inline demisto_sdk/commands/content_graph/interface/neo4j/queries/dependencies.py:60]: please use constant for DeprecatedContent and add a docstring

File: demisto_sdk/commands/content_graph/interface/neo4j/queries/dependencies.py [hunk 14]
```diff
@@ -104,6 +119,8 @@ def update_marketplaces_property(tx: Transaction, marketplace: str) -> None:
             "{marketplace}" IN content_item.marketplaces
         AND
             NOT "{marketplace}" IN dependency.marketplaces
+        AND
```
Comment (dantavori) [inline demisto_sdk/commands/content_graph/interface/neo4j/queries/dependencies.py:122]: add this to the docstring

File: demisto_sdk/commands/content_graph/interface/neo4j/queries/dependencies.py [hunk 15]
```diff
@@ -113,29 +130,72 @@ def update_marketplaces_property(tx: Transaction, marketplace: str) -> None:
...
+    WHERE {is_target_available("content_item", "integration")}
+    AND NOT command.object_id IN {list(GENERIC_COMMANDS_NAMES)}
+    WITH command, count(DISTINCT rcmd) as command_count
+
+    MATCH (content_item:{ContentType.BASE_CONTENT})
+        -[r:{RelationshipType.USES}]->
+            (command)<-[rcmd:{RelationshipType.HAS_COMMAND}]
+            -(integration:{ContentType.INTEGRATION})
     WHERE {is_target_available("content_item", "integration")}
+    AND NOT command.object_id IN {list(GENERIC_COMMANDS_NAMES)}
```
Comment (dantavori) [inline demisto_sdk/commands/content_graph/interface/neo4j/queries/dependencies.py:172]: same

File: demisto_sdk/commands/content_graph/interface/neo4j/queries/dependencies.py [hunk 16]
```diff
@@ -113,29 +130,72 @@ def update_marketplaces_property(tx: Transaction, marketplace: str) -> None:
...
 
-    WITH count(rcmd) as command_count, content_item, r, integration
     MERGE (content_item)-[u:USES]->(integration)
```
Comment (dantavori) [inline demisto_sdk/commands/content_graph/interface/neo4j/queries/dependencies.py:174]:
```suggestion
    MERGE (content_item)-[u:{RelationshipType.USES}]->(integration)
```

**Review by dantavori — APPROVED** [review]

**Pull request merged**